src/to_wiki.js

```diff
@@ -74,42 +76,132 @@ class WikiTextSerializerState {
+         nodeMarkLengths.set(node, lengths)
+      };
+
+      const children = []
+      node.forEach(c => children.push(c))
+      children.reverse()
+      children.forEach(findMarkLengths)
+
+      node.forEach(child => this.text(child, nodeMarkLengths))
+
+      this.end_inline()
+   }
 
-         // If openMarks includes marks that do not exist on the current node,
-         // close those marks before adding new marks and the inline text to
-         // the output.
-         let toClose = openMarks.filter(mark => marks.indexOf(mark) < 0)
-         this.closeMarks(toClose)
+   end_inline() {
+      // After all nodes have been handled, close any marks that are still open
+      this.closeMarks(this.currentlyOpenMarks)
+      this.currentlyOpenMarks = []
+      // Throw away spaces that were at the end of the text. We don't need them.
+      this.spaces = ''
+   }
 
-         // Remove closed marks from openMarks.
-         openMarks = openMarks.filter(mark => toClose.indexOf(mark) < 0)
+   /**
+    * Render a text node to the output
+    *
+    * If marklengths is available, it will be used to decide in what order to
+    * apply marks to the text
+    */
+   text(node, marklengths) {
+      let marks = node.marks.map(m => m.type.name)
 
-         // Marks that exist for the current node, but not in openMarks are new
-         // marks, so they should be opened.
-         let toOpen = marks.filter(mark => openMarks.indexOf(mark) < 0)
-         openMarks = toOpen.concat(openMarks)
+      let firstNeedle = function(haystack, needles) {
+         return needles.reduce((min, needle) =>
+          Math.min(min, haystack.indexOf(needle)), haystack.length)
+      }
 
-         this.openMarks(toOpen)
-         this.render(node)
+      // If openMarks includes marks that do not exist on the current node,
+      // close those marks before adding new marks and the inline text to
+      // the output.
+      const toClose = this.currentlyOpenMarks.
+       filter(openMark => marks.indexOf(openMark) < 0)
+      // We need to close all marks that were opened after the oldest mark we
+      // need to close, so that we don't get overlapping marks (i.e. `<i>italic
+      // <b>italic bold</i> bold</b>`, but in wiki text).
+      const earliestClose = firstNeedle(this.currentlyOpenMarks, toClose)
+
+      this.closeMarks(this.currentlyOpenMarks.slice(earliestClose))
+      // Remove closed marks from openMarks.
+      this.currentlyOpenMarks = this.currentlyOpenMarks.slice(0, earliestClose)
+
+      // Marks that exist for the current node, but not in currentlyOpenMarks
+      // are new marks, so they should be opened. This will include marks that
+      // were closed to get to the earliest close mark, since they won't be in
+      // currentlyOpenMarks.
+      const toOpen = []
+      marks.forEach(mark => {
+         if (this.currentlyOpenMarks.indexOf(mark) < 0) {
+            toOpen.push(mark)
+         }
+      })
```

Comment:
This is mostly a stylistic preference, but this could also be done with `filter`.

```js
const toOpen = marks.filter(mark => this.currentlyOpenMarks.indexOf(mark) < 0)
```